---
title: "ASURATDB"
author: "Keita Iida"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    toc: true
    toc_depth: 2
    number_section: true
    toc_float: TRUE
vignette: >
  %\VignetteIndexEntry{ASURATDB}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



# Installations
Attach necessary libraries:
```{r, message = FALSE, warning = FALSE}
library(ASURATDB)
```



<br>

# Collect Disease Ontology database for human cells
```{r, eval = FALSE}
library(DOSE) # For using `data(DO2EG)`
```

ASURATDB function `format_DO()` reformats a Disease Ontology database.
```{r, eval = FALSE}
data(DO2EG)
dict_DO <- enrichDO(unlist(DO2EG), ont = "DO", pvalueCutoff = 1,
                    pAdjustMethod = "BH", minGSSize = 0, maxGSSize = 1e+10,
                    qvalueCutoff = 1, readable = FALSE)
human_DO <- format_DO(dict = dict_DO@result, all_geneIDs = dict_DO@gene,
                      orgdb = org.Hs.eg.db::org.Hs.eg.db)
```



<br>

# Collect Cell Ontology database for human and mouse cells
ASURATDB functions `collect_CO()` and `format_CO()` load a Cell Ontology
database using ontoProc package and reformat the database, respectively.

**Tips:**
As of December 2020, Cell Ontology database might not be complete enough for
some biological contexts.
For example, well-known marker genes for pancreatic beta cell, Ins1 and Ins2,
were not registered for "type B pancreatic cell" with ID "CL:0000169".
```{r, eval = FALSE}
# Human
dict_CO <- collect_CO(orgdb = org.Hs.eg.db::org.Hs.eg.db)
human_CO <- format_CO(dict = dict_CO, orgdb = org.Hs.eg.db::org.Hs.eg.db)

# Mouse
dict_CO <- collect_CO(orgdb = org.Mm.eg.db::org.Mm.eg.db)
mouse_CO <- format_CO(dict = dict_CO, orgdb = org.Mm.eg.db::org.Mm.eg.db)
```



<br>

# Collect Gene Ontology database for human and mouse cells
ASURATDB functions `collect_GO()` and `format_GO()` load a Gene Ontology
database using clusterProfiler package and reformat the database, respectively.
Currently, only human and mouse data are acceptable.
```{r, eval = FALSE}
# Human
dict_GO <- collect_GO(orgdb = org.Hs.eg.db::org.Hs.eg.db)
human_GO <- format_GO(dict = dict_GO, orgdb = org.Hs.eg.db::org.Hs.eg.db)

# Mouse
dict_GO <- collect_GO(orgdb = org.Mm.eg.db::org.Mm.eg.db)
mouse_GO <- format_GO(dict = dict_GO, orgdb = org.Mm.eg.db::org.Mm.eg.db)
```



<br>

# Collect KEGG database for human and mouse cells
ASURATDB functions `collect_KEGG()` and `format_KEGG()` load a KEGG database
using KEGGREST package via the internet and reformat the database, respectively.

The arguments of `collect_KEGG()` are `organism` and `categories`.
Here, `organism` must obey the naming rule of
[KEGG](http://rest.kegg.jp/list/organism) and `categories` must be one of
`"pathway"` and `"module"` in the current version
(see `KEGGREST` function `listDatabases()`).
```{r, eval = FALSE}
# Human
dict_KEGG <- collect_KEGG(organism = "hsa", categories = c("pathway"))
human_KEGG <- format_KEGG(dict = list(pathway = dict_KEGG[["pathway"]][["success"]]),
                          orgdb = org.Hs.eg.db::org.Hs.eg.db)

# Mouse
dict_KEGG <- collect_KEGG(organism = "mmu", categories = c("pathway"))
mouse_KEGG <- format_KEGG(dict = list(pathway = dict_KEGG[["pathway"]][["success"]]),
                          orgdb = org.Mm.eg.db::org.Mm.eg.db)
```
Note `collect_KEGG()` uses `KEGGREST` function `keggGet()`,
which may produce both successful and unsuccessful results.



<br>

# Collect MSigDB for human cells
Load databases.
```{r, eval = FALSE}
dbtable <- clustermole::clustermole_markers()
```

```{r, eval = FALSE}
sort(unique(dbtable$db))
```
```
[1] "ARCHS4"     "CellMarker" "MSigDB"     "PanglaoDB"  "SaVanT"     "TISSUES"   
[7] "xCell"
```

Select species and databases.
```{r, eval = FALSE}
dbtable <- dbtable[which(dbtable$species == "Human"), ]
dbtable <- dbtable[which(dbtable$db == "MSigDB"),]
dbtable$geneID <- NA
```

Change gene symbols into entrez IDs.
```{r, eval = FALSE}
dictionary <- AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db,
                                    key = dbtable$gene_original,
                                    columns = c("SYMBOL", "ENTREZID"),
                                    keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
dictionary <- dictionary[which(!is.na(dictionary$SYMBOL)),]
for(i in 1:nrow(dbtable)){
  gene <- dbtable$gene_original[i]
  inds <- which(dictionary$SYMBOL == gene)
  dbtable$geneID[i] <- dictionary[inds,]$ENTREZID
}
```

Reformat the database.
Here, the identifier of each biological term are named "MSigID."
```{r, eval = FALSE}
dbtable_geneID <- split(x = dbtable$geneID, f = dbtable$celltype)
dbtable_symbol <- split(x = dbtable$gene_original, f = dbtable$celltype)
stopifnot(identical(length(dbtable_geneID), length(dbtable_symbol)))

res <- c("ID", "Description", "Count", "Gene", "GeneID", "IC")
res <- data.frame(matrix(ncol = 6, nrow = 0, dimnames = list(NULL, res)))
for(i in 1:length(dbtable_geneID)){
  res <- rbind(res, data.frame(
    ID = paste("MSigID:", i, sep = ""),
    Description = names(dbtable_geneID)[i],
    IC = NA,
    Count = length(dbtable_geneID[[i]]),
    Gene = paste(dbtable_symbol[[i]], collapse = "/"),
    GeneID = paste(dbtable_geneID[[i]], collapse = "/")))
}
human_MSig <- list(cell = res)
```



<br>

# Create a custom-built database
## Combine CO and MSigDB for human cells
Create a cell type-related database by combining Cell ontology and
MSigDB databases for analyzing human single-cell transcriptome data.
```{r, eval = FALSE}
urlpath <- "https://github.com/keita-iida/ASURATDB/blob/main/genes2bioterm/"
load(url(paste0(urlpath, "20201213_human_CO.rda?raw=true")))
load(url(paste0(urlpath, "20220108_human_MSig.rda?raw=true")))
res <- rbind(human_CO[["cell"]], human_MSig[["cell"]])
res$IC <- NA
human_COMSig <- list(cell = res)
```



<br>

## Combine DO, CO, and MSigDB for human cells
Create a cell type-related database by combining Disease Ontology,
Cell ontology and MSigDB databases for analyzing complex human single-cell
transcriptome data.
```{r, eval = FALSE}
urlpath <- "https://github.com/keita-iida/ASURATDB/blob/main/genes2bioterm/"
load(url(paste0(urlpath, "20201213_human_DO.rda?raw=true")))
load(url(paste0(urlpath, "20201213_human_CO.rda?raw=true")))
load(url(paste0(urlpath, "20220108_human_MSig.rda?raw=true")))
res <- rbind(human_DO[["disease"]], rbind(human_CO[["cell"]], human_MSig[["cell"]]))
res$IC <- NA
human_DOCOMSig <- list(cell = res)
```



<br>

# Session information

```{r}
sessionInfo()
```
